### PR TITLE
Operator overloading

### PR DESCRIPTION
**Pull request by tavisca-akash**
Please Review updated  changes @neilunadkat 
@ytokas

### CODE REVIEW

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+using OperatorOverloading.Model;
+namespace OperatorOverloading.Host
+{
+    class Program
+    {
+        public static void Main(string[] args)
+        {
+
+
+            Money monyObject = new Money();
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: Spell check

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            
+            //Take input Frst ammount
+            Console.WriteLine("Enter Ammount and currency");
+           
+            //split given line
+
+           String inputLine =(Console.ReadLine());
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: Why String and not string?

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+           string[] inputSplit = inputLine.Split(' ');
+           monyObject.Ammount = Convert.ToDouble(inputSplit[0]);
+           monyObject.Currency = inputSplit[1];
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: check if array has two values

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+            Money monyObject = new Money();
+            
+            //Take input Frst ammount
+            Console.WriteLine("Enter Ammount and currency");
+           
+            //split given line
+
+           String inputLine =(Console.ReadLine());
+           string[] inputSplit = inputLine.Split(' ');
+           monyObject.Ammount = Convert.ToDouble(inputSplit[0]);
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: use double.tryparse

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+           monyObject.Currency = inputSplit[1];
+            
+            Money moneyObject1 = new Money(monyObject.Ammount, monyObject.Currency);
+
+            ////Take input Frst ammount
+            Console.WriteLine("Enter Ammount and currency to add");
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: extract input parsing to separate method

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace OperatorOverloading.Model
+{
+    public class Money
+    {
+
+        double ammount;
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: private var starts with '_'

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        string currency;
+
+        //constructor to assign values
+        public Money(double real, string imaginary)
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: var names should be meaningful, should tell what they represent

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+        {
+            this.ammount = real;
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: Don't set private var.
this.Amount = value; not this.amount

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+            this.currency = imaginary;
+        }
+
+        //properties 
+        public double Ammount
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: no need for private var. 
public double Amount {get; set;}

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+        }
+
+        //constructor just to initalize object
+        public Money()
+        {
+
+        }
+
+        //operator overloading  
+        public static Money operator +(Money m1, Money m2)
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: variable names please

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+        {
+
+            if (m1.currency != m2.currency)
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: use string.Equals, and check for case sensitivity

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                return new Money(0,"Msmatch currency");
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: throw exception

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+            }
+            if (m2.ammount<0)
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: amount can be negative, no need for this check

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+            {
+                return new Money(0, "Negatve number can not be added");
+            }
```
Comment (tavisca-akash) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs:58]: Add check for overflow

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        {
+
+            double amountTemporary;
+            try
+            {
+                Money moneyObject = new Money();
+
+                //Take input First Amount
+                Console.WriteLine("Enter Amount");
+                if (Double.TryParse(Console.ReadLine(), out amountTemporary))
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: What if this is false? Add handling for the incorrect input.
Why Have you used Double and not double.

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+                    moneyObject.Amount = amountTemporary;
+
+                moneyObject.Currency = (Console.ReadLine());
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: Add null and empty checks for Currency.
Also check string length should be 3 char

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+                //Take input First Amount
+                Console.WriteLine("Enter Amount");
+                if (Double.TryParse(Console.ReadLine(), out amountTemporary))
+                    moneyObject.Amount = amountTemporary;
+
+                moneyObject.Currency = (Console.ReadLine());      
+                Money moneyObject1 = new Money(moneyObject.Amount, moneyObject.Currency);
+
+                ////Take input second Amount
+                moneyObject.Amount = Convert.ToDouble(Console.ReadLine());
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: use double.TryParse

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+        public Money()
+        {
+
+        }
+
+        //operator overloading  
+        public static Money operator +(Money money1, Money money2)
+        {
+
+            if (!string.Equals( money1.Currency,money2.Currency,StringComparison.CurrentCultureIgnoreCase))
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: Add null checks for money objects and their currencies. 
Why have you used CurrentCultureIgnoreCase?

**Comment by tavisca-akash** [comment]
@ytokas Review My final Code

File: 01 OperatorOverloading/OperatorOverloading.Host/Program.cs [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+            string amountTemporary;
+            string currencyTemporary;
+            try
+            {
+                Money moneyObject = new Money();
+
+                //Take input First Amount
+
+                Console.WriteLine("Enter Amount");
+                moneyObject.Amount =programObject.ValidateInputAmount();
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Host/Program.cs]: why do you need to make programObject to access the functions. You can directly  access them

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+            get;set;
+        }
+        
+
+        //constructor just to initalize object
+    //operator overloading  
+        public static Money operator +(Money money1, Money money2)
+        {
+            //If Object s null
+            if(money1.Equals("")||money2.Equals(""))
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs]: what is this? money1 == null

File: 01 OperatorOverloading/OperatorOverloading.Model/Money.cs [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+            {
+                throw new ArgumentNullException(Messages.NullArgument);
+            }
+
+            if (String.Equals(money1.Currency, money2.Currency, StringComparison.OrdinalIgnoreCase) == false)
+            {
+                throw new Exception(Messages.CurrencyMismatch);
+            }
+            
+            return new Money(money1.Amount + money2.Amount, money1.Currency);
```
Comment (ytokas) [inline 01 OperatorOverloading/OperatorOverloading.Model/Money.cs:61]: Where is amount overflow check.
Where is currency validations

File: Training/Program.cs [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using OperatorOverloading.Model;
+namespace OperatorOverloading.Host
+{
+    class Program
+    {
+        public static void Main(string[] args)
```
Comment (ytokas) [inline Training/Program.cs]: why are there two main functions. which one is the correct one

**Comment by tavisca-akash** [comment]
Please Review@ytokas

**Comment by tavisca-akash** [comment]
Please Review@ytokas

**Pull request merged**